microfit/histogram/histogram.py

```diff
@@ -417,8 +417,25 @@ def add_covariance(self, cov_mat):
         ----------
         cov_mat : array_like
             Covariance matrix to be added to the histogram.
+        fractional: bool
+            Whether passed covariance matrix is fractional or not
         """
 
+        # TODO: Check dimensions of passed covariance matrix are compatible with
+        # the hitsograms
+
+        if fractional:
+            if self.bin_counts.ndim == 1:
+                values = self.bin_counts
+            elif self.bin_counts.ndim == 2:
+                values = np.diag(self.bin_counts)
+            else:
+                raise ValueError("histogram bin counts can only be a 1D or 2D array")
+
+            for i in (0,len(cov_mat)-1):
+                for j in (0,len(cov_mat[i])-1):
+                    cov_mat[i][j] *= values[i]*values[j] 
```

Comment:
Use numpy array operations here instead of for-loops for better performance. This is an operation that we potentially do very often, even on cached histograms.